fv3core/stencils/nh_p_grad.py

```diff
@@ -90,60 +86,129 @@ def CalcV(
-    a2b_ord4.compute(delp, wk1)
-
-    CalcWk(pk3, wk, origin=orig, domain=(grid.nic + 1, grid.njc + 1, grid.npz))
-
-    du = utils.make_storage_from_shape(u.shape, origin=orig, cache_key="nh_p_grad_du")
-
-    CalcU(
-        u,
-        du,
-        wk,
-        wk1,
-        gz,
-        pk3,
-        pp,
-        grid.rdx,
-        dt,
-        origin=orig,
-        domain=(grid.nic, grid.njc + 1, grid.npz),
-    )
-
-    dv = utils.make_storage_from_shape(v.shape, origin=orig, cache_key="nh_p_grad_dv")
-
-    CalcV(
-        v,
-        dv,
-        wk,
-        wk1,
-        gz,
-        pk3,
-        pp,
-        grid.rdy,
-        dt,
-        origin=orig,
-        domain=(grid.nic + 1, grid.njc, grid.npz),
-    )
-    return u, v, pp, gz, pk3, delp
+
+    def __init__(self):
+        grid = spec.grid
+        self.orig = (grid.is_, grid.js, 0)
+        self.domain_full_k = (grid.nic + 1, grid.njc + 1, grid.npz)
+        self.domain_k1 = (grid.nic + 1, grid.njc + 1, 1)
+        self.u_domain = (grid.nic, grid.njc + 1, grid.npz)
+        self.v_domain = (grid.nic + 1, grid.njc, grid.npz)
+        self.nk = grid.npz
+        self.rdx = grid.rdx
+        self.rdy = grid.rdy
+
+        self._tmp_wk = utils.make_storage_from_shape(
+            grid.domain_shape_full(add=(0, 0, 1)), origin=self.orig
+        )  # pk3.shape
+        self._tmp_wk1 = utils.make_storage_from_shape(
+            grid.domain_shape_full(add=(0, 0, 1)), origin=self.orig
+        )  # pp.shape
+        self._tmp_du = utils.make_storage_from_shape(
+            grid.domain_shape_full(add=(0, 1, 0)), origin=self.orig
+        )
+        self._tmp_dv = utils.make_storage_from_shape(
+            grid.domain_shape_full(add=(1, 0, 0)), origin=self.orig
+        )
+
+        self._set_k0_stencil = stencil(
+            definition=set_k0,
+            backend=global_config.get_backend(),
+            rebuild=global_config.get_rebuild(),
+        )
+
+        self._calc_wk_stencil = stencil(
+            definition=CalcWk,
+            backend=global_config.get_backend(),
+            rebuild=global_config.get_rebuild(),
+        )
+
+        self._calc_u_stencil = stencil(
+            definition=CalcU,
+            backend=global_config.get_backend(),
+            rebuild=global_config.get_rebuild(),
+        )
+
+        self._calc_v_stencil = stencil(
+            definition=CalcV,
+            backend=global_config.get_backend(),
+            rebuild=global_config.get_rebuild(),
+        )
+
+    def __call__(
+        self,
+        u: FloatField,
+        v: FloatField,
+        pp: FloatField,
+        gz: FloatField,
+        pk3: FloatField,
+        delp: FloatField,
+        dt: float,
+        ptop: float,
+        akap: float,
+    ):
+        """
+        Updates the U and V winds due to pressure gradients,
+        accounting for both the hydrostatic and nonhydrostatic contributions.
+        Args:
```

Comment:
should we split between input and inout args to get a sense of data flow?

Reply:
Could do. Even though we return u, v, pp, gz, pk3, and delp I think only u and v are updated, so we could maybe stop returning the other variables